src/module.c

```diff
@@ -7388,6 +7405,262 @@ RedisModuleBlockedClient *moduleBlockClient(RedisModuleCtx *ctx, RedisModuleCmdF
+ * (1) Return REDISMODULE_AUTH_SUCCEEDED after successfully using the RM_Authenticate* APIs. This
+ * will immediately end the auth chain as successful and add the OK reply.
+ * (2) Return REDISMODULE_AUTH_DENIED after optionally setting `err` with an error message. This
+ * will immediately end the auth chain as unsuccessful and add the ERR reply.
+ * (3) Return REDISMODULE_AUTH_NOT_HANDLED. This will allow the engine to attempt the next custom
+ * auth callback.
+ * (4) Return REDISMODULE_AUTH_BLOCKED after using the RM_BlockClientOnAuth API. Here, the client
+ * will be blocked until the RM_UnblockClient API is used which will trigger the auth reply callback.
+ * In this reply callback, the Module should return one of the 3 above codes after following the
+ * steps mentioned.
+ * If none of the callbacks authenticate or deny auth, then password based auth is attemmpted
+ * and will authenticate or add failure logs and reply to the clients accordingly. */
+void RM_RegisterCustomAuthCallback(RedisModuleCtx *ctx, RedisModuleCustomAuthCallback cb) {
+    RedisModuleCustomAuthCtx *auth_ctx = zmalloc(sizeof(RedisModuleCustomAuthCtx));
+    auth_ctx->module = ctx->module;
+    auth_ctx->auth_cb = cb;
+    listAddNodeTail(moduleCustomAuthCallbacks, auth_ctx);
+}
+
+/* Returns 1 if the module has any custom auth cbs registered. If `should_unregister` is  non zero,
+ * all the module's custom auth cbs are unregistered.
+ * Returns 0 otherwise - if the Module has no custom auth cb registered. */
+int moduleSearchOrUnregisterCustomAuthCBs(RedisModule *module, int should_unregister) {
+    int callback_found = 0;
+    listIter li;
+    listNode *ln;
+    listRewind(moduleCustomAuthCallbacks, &li);
+    while ((ln = listNext(&li))) {
+        RedisModuleCustomAuthCtx *ctx = listNodeValue(ln);
+        if (ctx->module == module) {
+            callback_found = 1;
+            if (!should_unregister) return callback_found;
+            listDelNode(moduleCustomAuthCallbacks, ln);
+            zfree(ctx);
+        }
+    }
+    return callback_found;
+}
+
+/* Returns 1 if custom module based auth is in progress on any client. Returns 0 otherwise. */
+int isCustomAuthInProgress() {
+    if (!listLength(moduleCustomAuthCallbacks)) return 0;
+    listIter li;
+    listNode *ln;
+    listRewind(server.clients,&li);
+    while ((ln = listNext(&li)) != NULL) {
+        client *c = listNodeValue(ln);
+        if (c->flags & CLIENT_CUSTOM_AUTH) {
+            return 1;
+        }
+    }
+    return 0;
+}
+
+/* Unregister ALL the custom auth callbacks that this module had registered.
+ * Returns REDISMODULE_ERR in the following cases:
+ * - `ctx` is NULL.
+ * - ENOENT: The module does not have any custom auth cbs registered.
+ * - EINPROGRESS: Custom Auth is in progress and the module has a custom auth cb registered.
+ * Returns REDISMODULE_OK if the module had custom auth cbs and they were unregistered. */
+int RM_UnregisterCustomAuthCallbacks(RedisModuleCtx *ctx) {
```

Comment:
>  Other functionality such as configs lack this mirror API, since you have a strategy which is to just unload and reload the module
This makes sense - I agree that module unload can be used instead.

Reply:
It does seem redundant. I also incline with the above.

Reply:
Removed this API in - https://github.com/redis/redis/pull/11659/commits/5a234b6dc3afbdea1cb3c801bff113669aaf400c